odl/space/base_tensors.py

```diff
@@ -0,0 +1,646 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base classes for implementations of tensor spaces."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+from odl.set.sets import Set, RealNumbers, ComplexNumbers
+from odl.set.space import LinearSpace, LinearSpaceElement
+from odl.util.ufuncs import BaseGeneralizedTensorUfunc
+from odl.util.utility import (
+    is_scalar_dtype, is_floating_dtype, is_real_dtype,
+    arraynd_repr, dtype_str, repr_signature,
+    TYPE_MAP_R2C, TYPE_MAP_C2R)
+
+
+__all__ = ('BaseTensorSet', 'BaseTensorSpace')
+
+
+class BaseTensorSet(Set):
+
+    """Base class for sets of tensors of arbitrary type."""
+
+    def __init__(self, shape, dtype, order):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        shape : sequence of int
+            Number of elements per axis.
+        dtype :
+            Data type of each element. Can be provided in any
+            way the `numpy.dtype` function understands, e.g.
+            as built-in type or as a string.
+        order : {'C', 'F'}
+            Axis ordering of the data storage.
+        """
+        self.__shape = tuple(int(s) for s in shape)
+        if any(s < 0 for s in self.shape):
+            raise ValueError('`shape` must have only positive entries, got '
+                             '{}'.format(shape))
+        self.__dtype = np.dtype(dtype)
+        if self.ndim == 1:
+            self.__order = 'C'
+        else:
+            self.__order = str(order).upper()
+        if self.order not in ('C', 'F'):
+            raise ValueError("`order '' not understood".format(order))
+
+    @property
+    def shape(self):
+        """Number of elements per axis."""
+        return self.__shape
+
+    @property
+    def dtype(self):
+        """Data type of each tensor entry."""
+        return self.__dtype
+
+    @property
+    def order(self):
+        """Data storage order, either 'C' or 'F'."""
+        return self.__order
+
+    @property
+    def size(self):
+        """Total number of tensor entries."""
+        return np.prod(self.shape)
+
+    @property
+    def ndim(self):
+        """Number of dimensions (axes)."""
+        return len(self.shape)
+
+    def __len__(self):
+        """Number of tensor entries along the first axis."""
+        return self.shape[0]
+
+    @property
+    def itemsize(self):
+        """Size in bytes of one entry in an element of this space."""
+        return self.dtype.itemsize
+
+    @property
+    def nbytes(self):
+        """Number of bytes an element of this space uses in memory."""
```

Comment:
Wouldn't this make more sense as an element property rather than a space property?